skopt/utils.py

```diff
@@ -501,3 +499,104 @@ def normalize_dimensions(dimensions):
 
     return Space(transformed_dimensions)
+
+
+def wrap_named_args(dimensions):
+    """
+    Wrapper for an objective function that uses named arguments to make it
+    compatible with optimizers that use a single list of parameters.
+
+    Your objective function can be defined as being callable using named
+    arguments: `func(foo=123, bar=3.0, baz='hello')` for a search-space
+    with dimensions named `['foo', 'bar', 'baz']`. But the optimizer
+    will only pass a single list `x` of unnamed arguments when calling
+    the objective function: `func(x=[123, 3.0, 'hello'])`. This wrapper
+    converts your objective function with named arguments into one that
+    accepts a list as argument, while doing the conversion automatically.
+
+    The advantage of this is that you don't have to unpack the list of
+    arguments `x` yourself, which makes the code easier to read and
+    also reduces the risk of bugs if you change the number of dimensions
+    or their order in the search-space.
+
+    Example Usage
+    -------------
+    # Define the search-space dimensions. They must all have names!
+    dim1 = Real(name='foo', low=0.0, high=1.0)
+    dim2 = Real(name='bar', low=0.0, high=1.0)
+    dim3 = Real(name='baz', low=0.0, high=1.0)
+
+    # Gather the search-space dimensions in a list.
+    dimensions = [dim1, dim2, dim3]
+
+    # Define the objective function with named arguments
+    # and use this function-decorator to specify the search-space dimensions.
+    @wrap_named_args(dimensions=dimensions)
+    def my_objective_function(foo, bar, baz):
+        return foo ** 2 + bar ** 4 + baz ** 8
+
+    # Now the function is callable from the outside as
+    # `my_objective_function(x)` where `x` is a list of unnamed arguments,
+    # which then wraps your objective function that is callable as
+    # `my_objective_function(foo, bar, baz)`.
+    # The conversion from a list `x` to named parameters `foo`, `bar`, `baz`
+    # is done automatically.
+
+    # Run the optimizer on the wrapped objective function which is called as
+    # `my_objective_function(x)` as expected by `forest_minimize()`.
+    result = forest_minimize(func=my_objective_function, dimensions=dimensions,
+                             n_calls=20, base_estimator="ET", random_state=4)
+
+    # Print the best-found results.
+    print("Best fitness:", result.fun)
+    print("Best parameters:", result.x)
+
+    Parameters
+    ----------
+    * `dimensions` [list(Dimension)]:
+        List of `Dimension`-objects for the search-space dimensions.
+
+    Returns
+    -------
+    * `wrapped_func` [callable] 
+        Wrapped objective function.
+    """
+
+    def decorator(func):
+        """
+        This uses more advanced Python features to wrap `func` using a
+        function-decorator. To get explanations of how this works in Python,
+        please do an internet search for e.g.:
+        Python function decorator with additional arguments.
+        """
```

Comment:
I would just remove that, or make that comment shorter. This is standard practice in Python.

Reply:
This is an advanced usage of Python decorators. It is poorly understood even by experienced Python programmers as I could see on StackOverflow - I myself have never used this technique before and I had to look it up. It took me quite a while to find out how to do it, and it is not obvious where to find information on this. There is no good doc-page for this in the Python manuals. So to save people a lot of time, I give a good suggestion for a google search-string if they want to understand what is going on in this code.

Please stop being so goddamn stingy with comments in the code!

Reply:
> Please stop being so goddamn stingy with comments in the code!

I must admit, I laughed reading that.

Please read my comment again: I asked to either remove or shorten your comment. Telling people to google if they dont understand Python does not really help. 

So let's be constructive. Instead I suggest to point the readers to the page you read yourself to implement this pattern. This can be shortly expressed in 1 or 2 lines at most, and is more helpful to everyone.

Reply:
It was a bunch of pages and none of them were really good. I picked up a little info here and a little there. If you google for the 'wrong' keywords you find even worse pages. If there was a single good webpage explaining this I would of course have linked to it. I added these keywords because they really helped me. Again, this is not a common Python feature that is being used and even experienced developers on StackOverflow were confused about it.

Again you are deliberately nitpicking what you want changed in this PR.

You can bulldoze all these comments a month from now and I will never notice, but I suggest you instead have some empathy for who is going to be reading the code in the future.

Reply:
I agree with @glouppe. It would be better to state which pages exactly the user should visit where the helpful information is; Otherwise, every user who wants to understand the code will do the search from scratch, possibly ending on wrong pages. 

After some search, these two links seem to explain what is going on:

Explanation to `@wraps`:
 [https://docs.python.org/2/library/functools.html#functools.wraps](https://docs.python.org/2/library/functools.html#functools.wraps)

Explanation to decorator with parameters:
 [https://stackoverflow.com/questions/5929107/decorators-with-parameters](https://stackoverflow.com/questions/5929107/decorators-with-parameters)

Reply:
I concede and have rewritten these doc-strings. However, I didn't like the links you found as I found them confusing :-) I managed to find a 7 minute youtube video that explains it quite well, so I added that link. I hope everyone is happy with this.